File: Chapter 04/JS/section_4_4_anti_patterns/after_rich_domain_thin_controller/infrastructure/emailService.js
const { IEmailService } = require("../domain/interfaces/dataAccessInterfaces");

/**
 * Simulates sending an order confirmation email via SMTP.
 * @implements {IEmailService}
 */
class SmtpEmailService extends IEmailService {
    sendOrderConfirmation(order) {
        const customerName = order.getCustomer().name;
        const totalPrice = order.totalPrice;

        // In a real system, you would connect to an SMTP server
        // and send a formatted email.
        console.log("---");
        console.log("Connecting to SMTP server...");
        console.log(`Sending email to ${customerName}:`);
        console.log("Subject: Your Order Confirmation");
        console.log(`Body: Thank you for your order of $${totalPrice.toFixed(2)}.`);
        console.log("Email sent.");
        console.log("---");
    }
}

module.exports = { SmtpEmailService };

File: Chapter 04/Python/section_4_4_anti_patterns/after_rich_domain_thin_controller/infrastructure/email_service.py
from ..domain.interfaces.data_access_interfaces import IEmailService


class SmtpEmailService(IEmailService):
    """
    ARCHITECTURE NOTE: By isolating Email logic here, we prevent 
    database concerns from "leaking" into the Presentation or 
    Business layers.
    """
    # Concrete implementation for an email provider
    def send(self, to: str, subject: str, body: str):
        pass